### PR TITLE
feat: Git include exclude feature and partial commit search 

### PR DESCRIPTION
**Pull request by vikramdevtron**
# Description

While configuration git repository, users may provide you the include and exclude specific paths for builds, allowing only those materials for builds which are within these specific paths.
Filter pattern only applicable on branch fixed source type.

Fixes #https://github.com/devtron-labs/devtron/issues/3129

## How Has This Been Tested?
- [x] Fetch Ci materials will fetch after applying Filters
- [x] Fetch Ci materials with fetch after applying Filters, also include excluded items, with marked. 
- [x] Auto trigger will block for excluded materials
- [x] commit search can be found based on partial search also.

## Checklist:

* [x] The title of the PR states what changed and the related issues number (used for the release note).
* [x] Does this PR requires documentation updates?
* [ ] I've updated documentation as required by this PR.
* [ ] I have performed a self-review of my own code.
* [ ] I have commented my code, particularly in hard-to-understand areas.
* [x] I have tested it for all user roles.
* [ ] I have added all the required unit/api test cases.

## Does this PR introduce a user-facing change?

```release-note

```

### CODE REVIEW

**Review by iamayushm — APPROVED** [review]

File: pkg/RepoManages.go [hunk 1]
```diff
@@ -459,7 +459,26 @@ func (impl RepoManagerImpl) FetchGitCommitsForBranchFixPipeline(pipelineMaterial
...
 	}
-	response.Commits = commits
+	if len(gitMaterial.FilterPattern) == 0 {
+		response.Commits = commits
+		return response, nil
+	}
+
+	filterCommits := make([]*git.GitCommit, 0)
+	for _, commit := range commits {
+		impl.logger.Info("________________________________________________")
```
Comment (kripanshdevtron) [inline pkg/RepoManages.go:469]: do we need this log now ??

File: pkg/RepoManages.go [hunk 2]
```diff
@@ -459,7 +459,26 @@ func (impl RepoManagerImpl) FetchGitCommitsForBranchFixPipeline(pipelineMaterial
...
+		excluded := impl.gitWatcher.PathMatcher(commit.FileStats, gitMaterial)
+		impl.logger.Infow("include exclude result", "excluded", excluded)
```
Comment (kripanshdevtron) [inline pkg/RepoManages.go:471]: we should add metadata to it and also this should be in DEBUG mode

File: pkg/git/RepositoryManager.go [hunk 3]
```diff
@@ -259,6 +259,11 @@ func (impl RepositoryManagerImpl) ChangesSinceByRepository(repository *git.Repos
 			Date:    commit.Author.When,
 			Message: commit.Message,
 		}
+		stats, err := commit.Stats()
+		if err != nil {
+			impl.logger.Errorw("error in  fetching stats", "err", err)
```
Comment (kripanshdevtron) [inline pkg/git/RepositoryManager.go:264]: we should log repoName

File: pkg/git/Watcher.go [hunk 4]
```diff
@@ -291,6 +301,79 @@ func (impl GitWatcherImpl) SubscribeWebhookEvent() error {
...
+		path := fileChange["Name"].(string)
+		changesInPath = append(changesInPath, path)
+	}
+	impl.logger.Infow("pathMatcher .............", "changes in paths", changesInPath)
+	len := len(pathsForFilter)
+	for i, filter := range pathsForFilter {
+		isExcludeFilter := false
+		isMatched := false
+		//TODO - handle ! in file name with /!
+		if strings.Contains(filter, "!") {
```
Comment (kripanshdevtron) [inline pkg/git/Watcher.go:338]: we can make "!" a constant so that we also understands the intent of this mark.

**Review by kripanshdevtron — APPROVED** [review]

**Pull request merged**